enarx-keep-sev-shim/src/hostcall.rs

```diff
@@ -0,0 +1,208 @@
+        }
+    }
+
+    /// Acquire a HostCall instance to communicate with the host.
+    ///
+    /// # Caution
+    ///
+    /// Do not call this in interrupts, as it can cause deadlocks
+    #[inline(always)]
+    pub fn lock() -> Self {
+        loop {
+            // Try to take out the virtual address out of the global variable
+            let address = unsafe { HOSTCALL_VIRT_ADDR.swap(0, Ordering::Acquire) };
+            if address == 0 {
+                // Already in use
+                core::sync::atomic::spin_loop_hint();
+            } else {
+                return Self(address as *mut Block);
+            }
+        }
+    }
+
+    /// Causes a `#VMEXIT` for the host to process the data in the shared memory
+    ///
+    /// Returns the contents of the shared memory reply status, the host might have
+    /// written.
+    ///
+    /// # Safety
+    ///
+    /// The parameters returned can't be trusted.
+    #[inline(always)]
+    unsafe fn hostcall(self) -> Result<Register<usize>, libc::c_int> {
+        let mut port = Port::<u16>::new(SYSCALL_TRIGGER_PORT);
+        port.write(1 as u16);
+        (*self.0).msg.rep.into()
+    }
+
+    /// Write `bytes` to a host file descriptor `fd`
+    ///
+    /// # Panics
+    ///
+    /// Panics, if `bytes.len()` exceeds `Block::buf_capacity()`
+    ///
+    /// # Safety
+    ///
+    /// The parameters returned can't be trusted.
+    pub unsafe fn write(self, fd: usize, bytes: &[u8]) -> Result<libc::c_int, libc::c_int> {
+        let count = bytes.len();
+        assert!(count <= Block::buf_capacity());
+
+        (*self.0).msg.req.num = Register::from_raw(libc::SYS_write as usize);
+        (*self.0).msg.req.arg[0] = Register::from_raw(fd);
+        (*self.0).msg.req.arg[1] =
+            ShimVirtAddr::try_from(Address::from((*self.0).buf.as_mut_ptr()))
+                .map_err(|_| libc::EFAULT)?
+                .into_phys_address()
+                .into();
+        (*self.0).msg.req.arg[2] = Register::from_raw(count as _);
+        (*self.0).buf[..count].copy_from_slice(&bytes[..count]);
+        self.hostcall().map(|r| r.raw() as _)
+    }
+
+    /// Exit the shim with a `status` code
+    ///
+    /// # Panics
+    ///
+    /// Panics, if the shim resumes to run.
+    #[inline(always)]
+    pub fn exit_group(self, status: u32) -> ! {
+        unsafe {
+            (*self.0).msg.req.num = Register::from_raw(libc::SYS_exit_group as _);
+            (*self.0).msg.req.arg[0] = Register::from_raw(status as _);
+
+            let _ = self.hostcall();
+
+            unreachable!()
+        }
+    }
+}
+
+impl Drop for HostCall {
+    fn drop(&mut self) {
+        // Put back in the virtual address in the global variable
+        let prev = unsafe { HOSTCALL_VIRT_ADDR.swap(self.0 as _, Ordering::Release) };
+        assert_eq!(prev, 0);
+    }
+}
+
+/// The address of the unencrypted page used to communicate with the host
+static mut HOSTCALL_VIRT_ADDR: AtomicUsize = AtomicUsize::new(0);
```

Comment:
Please move this to the top of the module.

Reply:
done